Remove jsnullablet

idkjs opened the pull request:
Per https://github.com/Schniz/bs-faker/pull/42#issuecomment-565809527, changed getOrRaise to `Belt.Option.getExn`.  Removes some of the `Js.Nullable.t` calls in bs-platform@7.0.1. I went and checked and it works. Obviously, this is bs.platform@7.01 nit picking. This will make the previous merge fully idiomatic bs.platform@7.01. Tests are passing.

Schniz commented on src/Company.rei:
```diff
@@ -1,5 +1,5 @@
 let suffixes: unit => Js.Array.t(string);
-let companyName: (~format: int=?, unit) => string;
+let companyName: (~format: int=?, unit) => option(string);
```
Can `companyName` return a nullable result?

Schniz replied:
Because this is why we need to `getExn` 😄 I'm not sure that these changes are needed

This relates to every one of the changes in the function signatures

idkjs replied:
If i take out `Belt.Option.getExn` it throw this error:
```sh
Error: This expression has type Js.String.t => array(Js.String.t)
       but an expression was expected of type option(string) => 'a
       Type Js.String.t = string is not compatible with type option(string) 
```

How do we handle that if we don't need `getExn`?

Schniz replied:
It looks like a constraint that you added in this PR (making companyName return an option instead of a string)

Schniz replied:
Am I missing something?

idkjs replied:
You correct. I have removed it and tests pass

idkjs commented on src/Company.re:
```diff
@@ -1,8 +1,6 @@
 type t = {
   suffixes: unit => Js.Array.t(string),
-  companyName: Js.Nullable.t(int) => string,
+  companyName: int => string,
```
So, thanks for checking through this code. I am trying to go through what happened here. Basically I was leaning hard on the tests. Since they all passed, I figured we were good. I don't really understand test construction yet so I am wondering if you can tell me what error in thinking I have made that you are keying in on. Obviously, you are seeing something obvious that the tests are not. What is going on here? Thanks in advance for the lesson, @Schniz .

Schniz replied:
no problems, thanks for contributing :) 

The idea here is that tests don't need to change, because we already had a Reason/OCaml wrapper around Faker that transformed every `option('a)` to `Js.Nullable.t('a)`. So test files shouldn't be changed. The fact that we now need to handle `option`s using `Belt.Option.getExn` show that our Reason API changed which shouldn't happen in this PR.

Optional arguments have weird syntax to them, because of the curry-by-default nature of OCaml: https://reasonml.github.io/docs/en/function#optional-labeled-arguments

you can use labeled arguments and make them optional by providing a positional `unit` argument like so:

```reason
let my_fn = (~optional_arg=?, ()) => { /* ... */ };

my_fn(); // works!
my_fn(~optional_arg="hello!", ()); // pass the optional argument
my_fn(()); // works, this is the same as `my_fn();`
```

If you'll try to do something like the following:

```reason
let my_fn = (~optional_arg=?) => { /* ... */ };
```

[You'll get a warning](https://reasonml.github.io/en/try?rrjsx=true&reason=DYUwLgBAtgng+gMwHYQLwQBQD8D2AHMASxyQENg5SAnAc1QH4BKNAPggG8IB6AKggDpBEHlwgBfANxA), saying:

```
This optional parameter in final position will, in practice, not be optional.
Reorder the parameters so that at least one non-optional one is in final position or, if all parameters are optional, insert a final ().

Explanation: If the final parameter is optional, it'd be unclear whether a function application that omits it should be considered fully applied, or partially applied. Imagine writing `let title = display("hello!")`, only to realize `title` isn't your desired result, but a curried call that takes a final optional argument, e.g. `~showDate`.

Formal rule: an optional argument is considered intentionally omitted when the 1st positional (i.e. neither labeled nor optional) argument defined after it is passed in.
```

tl;dr - writing in Reason:

```reason
let my_fn = (~optional_arg=?, ()) => { /* ... */ };
```

is practically the same as [writing the following in TypeScript](https://www.typescriptlang.org/play/?ssl=5&ssc=3&pln=3&pc=1#code/C4TwDgpgBAzg9gWwgfVJKBeKA7ArggIwgCcBuAKHIBsJgoERkAzbTKACgG8o4xgBLONgCGVZMOIBzKAF8AXFG68BQ0eKkB+BfCSpw0GW04yAlJgB8i8lCgB6W1AB0z8jNJA):

```typescript
let my_fn = ({ optional_arg }: { optional_arg?: some_type } = {}) => {
  // ...
};
```

idkjs replied:
Wow. This is an excellent explanation. Thank you for taking the time to write this.

Schniz commented:
And I think now, with the recent changes, the changes in the tests can be reverted to make them pass again

idkjs commented on src/Address.rei:
```diff
@@ -7,10 +7,10 @@ let county: unit => string;
 let longitude: unit => string;
 let secondaryAddress: unit => string;
-let state: (~useAbbr: bool=?, unit) => string;
+let state: (~useAbbr: bool=?, unit) => option(string);
```
reverted

idkjs commented on __tests__/Company_test.re:
```diff
@@ -7,17 +7,17 @@ describe("BsFaker.Company", () => {
-    test("accepts optional args", () =>
-      expect(Js.typeof(Company.companyName())) === "string"
-    );
+    // test("accepts optional args", () =>
```
You are correct. I didn't realized that these calls were redundant. Since the incoming value was an option, i assumed it had to return an option. I made the recommended edits and it still compiles AND the tests pass.

idkjs commented:
I think we are good now.

Schniz commented:
Cool! does everything still works if we do `git checkout origin/master -- src/*.rei`? This PR still have changes in the interface files and I think that keeping them unchanged due to formatting (or maybe just removing `type nullable = Js.Nullable.t`) will have a better "git hygiene"

I'd be also happy to get the code reformatted with `bsrefmt --in-place src/*.re*`

idkjs commented:
@Schniz What doe `git checkout origin/master -- src/*.rei` do ?

Schniz commented:
revert all changes in interface files

idkjs commented:
I reverted but then removed the `[bs.deriving abstract]` and changed the `[bs.optional`'s to `option(string)` in `Fakers.re` `type card_address` since the point here was to remove those calls. Is this what you expected? I am assuming the point was to reformat the `.rei` files. Is this correct?

Schniz commented:
The reason I wanted the `*.rei` to be almost unchanged is because it means that we can bump a minor version and not a major version, because there is no change in the public API 😄
